common/src/main/java/com/wynntils/handlers/actionbar/ActionBarHandler.java

```diff
@@ -0,0 +1,113 @@
+ * This file is released under AGPLv3. See LICENSE for full license details.
+ */
+package com.wynntils.handlers.actionbar;
+
+import com.wynntils.core.WynntilsMod;
+import com.wynntils.core.components.Handler;
+import com.wynntils.mc.event.ChatPacketReceivedEvent;
+import com.wynntils.mc.objects.ChatType;
+import com.wynntils.wynn.utils.WynnUtils;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import net.minecraft.network.chat.Component;
+import net.minecraftforge.eventbus.api.SubscribeEvent;
+
+public final class ActionBarHandler extends Handler {
+    // example: "§c❤ 218/218§0    §7502§f S§7 -1580    §b✺ 1/119"
+    private static final Pattern ACTIONBAR_PATTERN = Pattern.compile("(?<LEFT>§[^§]+)(?<CENTER>.*)(?<RIGHT>§[^§]+)");
+    public static final String CENTER_PADDING = "§0               ";
+
+    private Map<ActionBarPosition, List<ActionBarSegment>> allSegments = Map.of(
+            ActionBarPosition.LEFT,
+            new ArrayList<>(),
+            ActionBarPosition.CENTER,
+            new ArrayList<>(),
+            ActionBarPosition.RIGHT,
+            new ArrayList<>());
+    private Map<ActionBarPosition, ActionBarSegment> lastSegments = new HashMap<>();
+    private String previousRawContent = null;
+    private String previousProcessedContent;
+
+    public void registerSegment(ActionBarSegment segment) {
+        allSegments.get(segment.getPosition()).add(segment);
+    }
+
+    public void unregisterSegment(ActionBarSegment segment) {
+        ActionBarPosition pos = segment.getPosition();
+        segment.removed();
+        allSegments.get(pos).remove(segment);
+        if (lastSegments.get(pos) == segment) {
+            lastSegments.remove(pos);
+        }
+    }
+
+    @SubscribeEvent
+    public void onActionBarUpdate(ChatPacketReceivedEvent event) {
+        if (!WynnUtils.onWorld() || event.getType() != ChatType.GAME_INFO) return;
+
+        String content = event.getMessage().getString();
+        if (content.equals(previousRawContent)) {
+            // No changes, skip parsing
+            if (!content.equals(previousProcessedContent)) {
+                event.setMessage(Component.literal(previousProcessedContent));
+            }
+            return;
+        }
+        previousRawContent = content;
+
+        Matcher matcher = ACTIONBAR_PATTERN.matcher(content);
+        if (!matcher.matches()) {
+            WynntilsMod.warn("ActionBarHandler pattern failed to match: " + content);
+            return;
+        }
+
+        // Create map of position -> matching part of the content
+        Map<ActionBarPosition, String> positionMatches = new HashMap<>();
+        Arrays.stream(ActionBarPosition.values()).forEach(pos -> positionMatches.put(pos, matcher.group(pos.name())));
+
+        Arrays.stream(ActionBarPosition.values()).forEach(pos -> {
```

Comment:
Kind of hard to get this, this code checks the 3 action bar parts, LEFT, CENTER, RIGHT and checks if there are one or more "segments" in it, which we parse? Can one "side" have multiple "segments" in it? Or, did I not get this at all?

Reply:
Yes, correctly. I did not know what to call things. Left is technically always health and right is always mana (or, actually, right can be bubbles when you are underwater, yes?). The center part is the weird thing. It can be coordinates, sprint meter, spell cast indicator, or powder special. At least these are what I know. I called these "segments". Maybe it should rather be "SegmentHandler" or something, I dunno, but now I kind of kidnapped "handler" to mean something different and more specifically.

Reply:
so we try to figure out which segment (if any) is matching which part, in practice only for center.

Reply:
Okay. Should be fine then.